design-proposals/host-ip-pool.md

```diff
@@ -0,0 +1,486 @@
+to pass Nmstate configuration that will be rendered into NetworkManager keyfiles.
+So it's really the IPI installer config networkData API.
+https://github.com/openshift/installer/blob/969378dc498336d1ac29baa89f75324a44467d5d/pkg/types/baremetal/platform.go#L43
+
+### Non-Goals
+
+- This design does not cover IP address configuration of management NIC on day-0/1
+
+## Proposal
+
+**cluster admin** is a user responsible for managing the cluster node
+networking.
+
+### Workflow Description (assign static address with IPPool)
+
+1. The cluster admin creates `NodeIPPool` resources with IP ranges and a name.
+2. The cluster admin configures node network interfaces statically, with IPs from the `NodeIPPool` using a
+`NodeNetworkConfigurationPolicy` referencing them.
+3. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed
+
+#### Variation (exhausted range at the NodeIPPool)
+
+When the IP pool does not have an available IP address, the `NNCP` will fail
+with a explicit message, the `NodeIPPool` has to increase pool size.
+
+Modifying `NodeIPPool` means increasing CIDR, changing begin/end or exclude,
+since allocations are stored as offsets they will have to be re-calculated from
+the full IP present at NNCEs.
+
+### Workflow Description (removing a reference to NodeIPPool)
+
+1. The cluster admin changes the network interface to remove the reference to the `NodeIPPool` (for example harcoding the IP or switching to DHCP).
+2. The cluster admin waits for the NodeNetworkConfigurationPolicy to succeed, freeing the IPs from those interfaces.
+
+### Workflow Description (scaling up cluster)
+
+1. The cluster admin adds a new node affected by the `NodeNetworkConfigurationPolicy`.
+2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, causing the new node to take the address from the `NodeIPPool`
+
+### Workflow Description (scaling down cluster)
+
+1. The cluster admin removes a node that was taking an address from a `NodeIPPool`.
+2. The cluster admin waits for the `NodeNetworkConfigurationPolicy` to succeed, thus freeing the IPs from the removed node.
+
+### Workflow Description (manual NodeIPPool clean up)
+
+1. The cluster admin finds some stale IPs, They are inconsistencies between `NodeIPPool.status` and `NodeNetworkConfigurationEnactment.status`.
+2. The cluster admin resets the `NodeIPPool` allocated status specifying an empty list `status.allocated-offsets: []`.
+3. The cluster admin waits for `NodeIPPool` status to succeed
+
+NOTE: This is a manual garbage collector that is good enough for first implementation it can be converted into automatic in the future.
+
+### API Extensions day 2
+
+This proposal adds a new CRD to configure and store IP pools (NodeIPPool) and some
+changes at the NNCPs to reference the ippool.
+
+Following is an example CR for the cluster wide CRD `NodeIPPool`:
+
+```yaml
+apiVersion: nmstate.io/v1
+kind: NodeIPPool
+metadata:
+  name: traffic-1
+spec:
+  range:
```

Comment:
Why would a pool provide addresses from multiple ranges @AlonaKaplan ? 

Assuming there are multiple ranges, If a node requests an address from a pool, from which of the ranges will it get the IP ? The first ? Round-robin ? Random ?...

Reply:
yes, randomly one of the ranges.
Regarding the usage, I"m not sure. But for example in the [cisco documentation](https://www.cisco.com/c/en/us/td/docs/net_mgmt/prime/network/4-2-3/user/guide/CiscoPrimeNetwork423UserGuide/ip-pool.pdf) there is a concept of assigning an ip address from multiple ranges (group of pools).

Reply:
I don't know ... Let's say you have a range with ipv6, and another with ipv4. 

This would leave you with either an ipv4 or ipv6 address, depending on... randomness.

If we take this esoteric route (I find it strange) I would at least assure that all ranges belong to the same family.

Reply:
I thought about this a bit more and I honestly don't understand why would a pool feature multiple ranges. It can, but, the same can be done with multiple pools. 

Plus, I think if you have multiple ranges, you'll also have to keep track on those allocations for the different ranges in the pool. I just don't see the motivation for it - thus, I wouldn't account for it **until** there's a use case that requires it.

Reply:
@AlonaKaplan since every range will need it's own allocated info it does not gives any value define multiple of them at the same `NodeIPPool` we are not going to have a massive number of ranges when we talk about interfaces on nodes.